samples/gibbs_ensemble_client.py

```diff
@@ -0,0 +1,195 @@
+MSG_EXCHANGE_PART_ADD_REVERT = 41
+MSG_EXCHANGE_PART_REMOVE = 5
+MSG_EXCHANGE_PART_REMOVE_REVERT = 51
+MSG_ENEGRY = 6
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-n', '--number-particles', type=int, nargs=1)
+parser.add_argument('-lj', '--lj-params', type=float, nargs=4)
+parser.add_argument('-s', '--seed', type=int, nargs=1)
+parser.add_argument('-bl', '--box-length', type=float, nargs=1)
+
+args = parser.parse_args()
+
+if args.number_particles:
+    particle_number = args.number_particles[0]
+if args.lj_params:
+    lj_epsilon = args.lj_params[0]
+    lj_sigma = args.lj_params[1]
+    lj_cutoff = args.lj_params[2]
+    lj_shift = args.lj_params[3]
+if args.seed:
+    seed = args.seed[0]
+if args.box_length:
+    init_box_l = args.box_length[0]
+
+
+class Manipulator():
+    '''
+    Manipulator class, which has all the system changeing functions.
+    '''
+
+    def __init__(self, system):
+        self.m_system = system
+        self.m_old_part_position = []
+        self.m_old_part_idx = None
+
+    def move_particle(self):
+        '''Moves a particle to a new random position. The old position is saved in m_old_part_position.'''
+        sel = self.m_system.part.select(lambda p: p.id > -1).id_selection
+        self.m_old_part_idx = np.random.choice(sel)
+        self.m_old_part_position = self.m_system.part[self.m_old_part_idx].pos
+        self.m_system.part[self.m_old_part_idx].pos = np.random.rand(3) * box_l
+
+    def move_particle_revert(self):
+        '''Revert the last movement step.'''
+        self.m_system.part[self.m_old_part_idx].pos = self.m_old_part_position
+
+    def remove_particle(self):
+        '''Removes a random particle. The old position and index are saved in m_old_part_position and m_old_part_idx.'''
+        sel = self.m_system.part.select(lambda p: p.id > -1).id_selection
+        self.m_old_part_idx = np.random.choice(sel)
+        self.m_old_part_position = self.m_system.part[self.m_old_part_idx].pos
+        self.m_system.part[self.m_old_part_idx].remove()
+
+    def remove_particle_revert(self):
+        '''Revert the last remove particle step.'''
+        self.m_system.part.add(pos=self.m_old_part_position)
+
+    def change_volume_and_rescale_particles(self, new_box_l):
+        '''Changes the volume and rescales the particle positions.'''
+        self.m_system.change_volume_and_rescale_particles(new_box_l)
+
+    def insert_particle(self):
+        '''Inserts a particle at a random position.'''
+        self.m_system.part.add(pos=np.random.rand(3) * self.m_system.box_l)
+
+    def remove_last_added_particle(self):
+        '''Removes the last added particle (reverts intert particle).'''
+        self.m_system.part[self.m_system.part.highest_particle_id].remove()
+
+
+def recv_data():
+    '''Receives data and return it.'''
+    # The first 4 bytes encode the length of the messages received.
+    buf = b''
+    while len(buf) < 4:
+        buf += socket.recv(4 - len(buf))
```

Comment:
there are still global variables involved here. Please try to extract them as parameters to the function